Resolve cases when border and padding are added twice to computed min and max sizes

sammygill opened the pull request:
#### 0599c10f90042d00785952e18f8897082d655ca3
<pre>
Resolve cases when border and padding are added twice to computed min and max sizes
<a href="https://bugs.webkit.org/show_bug.cgi?id=243887">https://bugs.webkit.org/show_bug.cgi?id=243887</a>
rdar://98577125

Reviewed by Rob Buis.

This patch fixes a miscellaneous bug which would result in the
border and padding on boxes to get added twice in certain cases. This
is because inlineSizeFromAspectRatio returns the sizes of the border
box when the box sizing is content box. This is what is used to
compute logicalMinWidth and logicalMaxWidth, so those become
border box values. If the border and padding is not removed it is
possible for m_minPreferredLogicalWidth or m_maxPreferredLogicalWidth
to get clamped to one of those values. At the end of the method the
border and padding would once again get added on, which would result
in an incorrect box size. This could also happen in
RenderBlock::computePreferredLogicalWidths if the width is calculated
from the aspect ratio. This is because this branch of the code uses the
same inlineSizeFromAspectRatio method to do so.

Some of the logic inside of RenderBox::computePreferredLogicalWidths has
also been adjusted to no longer bother with checking transferred sizes
if a definite width is provided. This is because m_minPreferredLogicalWidth
and m_maxPreferredLogicalWidth will be set to the defined logical width
in RenderBlock::computePreferredLogicalWidths. If this is done, then
the transferred sizes will have no impact on these values. This is
because:

1.) If m_minPreferredLogicalWidth is greater than the logical width or
m_maxPreferredLogicalWidth is smaller than the logical width, they would
have been sized to the logical width anyway
2.) The clamping will have no effect if 1 does or does not occur. If 1
does occur, then m_minPreferredLogicalWidth and/or m_maxPreferredLogicalWidth
will already be the value of the logical width. If 1 does not ocur, then
m_minPreferredLogicalWidth and m_maxPreferredLogicalWidth are already
within the bounds of the transferred sizes and nothing will happen.

Transferred sizes spec: <a href="https://drafts.csswg.org/css-sizing-4/#aspect-ratio-size-transfers">https://drafts.csswg.org/css-sizing-4/#aspect-ratio-size-transfers</a>

* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computePreferredLogicalWidths):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computePreferredLogicalWidths):

Canonical link: <a href="https://commits.webkit.org/254190@main">https://commits.webkit.org/254190@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f89c60eaa4315f28e88f43dc2c4038d6261e0133

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88075 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32239 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18789 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97217 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/152711 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/92043 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30606 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26545 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80170 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91956 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93685 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24667 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74718 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24633 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79589 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79711 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67617 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/28235 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13600 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/28333 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14587 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2947 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/31361 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37475 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/30309 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33822 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -2308,9 +2308,11 @@ void RenderBlock::computePreferredLogicalWidths()
-    else
+    else if (shouldComputeLogicalWidthFromAspectRatio()) {
+        m_minPreferredLogicalWidth = m_maxPreferredLogicalWidth = (computeLogicalWidthFromAspectRatio() - borderAndPaddingLogicalWidth());
+        m_minPreferredLogicalWidth = std::max(LayoutUnit(0), m_minPreferredLogicalWidth);
```
I think using 0_lu is preferred these days.

sammygill replied:
Changed!

rwlbuis reviewed (changes requested):
It would be good to have some spec links in the commit message.

cathiechen commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -3553,6 +3553,16 @@ void RenderBox::computePreferredLogicalWidths(const Length& minWidth, const Leng
+        logicalMinWidth = std::max(logicalMinWidth - borderAndPadding, LayoutUnit(0));
+        logicalMaxWidth = std::max(logicalMaxWidth - borderAndPadding, LayoutUnit(0));
+
+        if (style().width().isFixed()) {
```
Use `style().logicalWidth()` instead of `style().width()` here? And if `style().logicalWidth().isFixed()`, there is no need to calculate transferred min/max width, so we could move it to the `if` constraint.
We have `RenderBox::constrainLogicalMinMaxSizesByAspectRatio` to handle different scenarios of transferred min/max size. Maybe in the follow-up patch, we can try to reuse this function here?

sammygill replied:
Changed! And yes that makes sense. There are still some flex box WPT tests that are failing so I can try to add that along with one of those patches.

cathiechen commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -3563,6 +3563,16 @@ void RenderBox::computePreferredLogicalWidths(const Length& minWidth, const Leng
 {
     if (shouldComputeLogicalHeightFromAspectRatio()) {
```
Sorry, I didn't express clearly! I mean, if `style().logicalWidth().isFixed()`, there is no need to check aspect-ratio.
It will be something like this:
```
    if (!style().logicalWidth().isFixed() && shouldComputeLogicalHeightFromAspectRatio()) {
        auto [logicalMinWidth, logicalMaxWidth] = computeMinMaxLogicalWidthFromAspectRatio();
        logicalMinWidth = std::max(logicalMinWidth - borderAndPadding, 0_lu);
        logicalMaxWidth = std::max(logicalMaxWidth - borderAndPadding, 0_lu);
        m_minPreferredLogicalWidth = std::clamp(m_minPreferredLogicalWidth, logicalMinWidth, logicalMaxWidth);
        m_maxPreferredLogicalWidth = std::clamp(m_maxPreferredLogicalWidth, logicalMinWidth, logicalMaxWidth);
    }
```

sammygill replied:
Is there a reason we don't want to check the aspect-ratio in that scenario? Would that not be this part of the spec: 

"This transferred minimum is capped by any definite [preferred](https://drafts.csswg.org/css-sizing-3/#preferred-size) or [maximum size](https://drafts.csswg.org/css-sizing-3/#max-width) in the destination axis." 

I believe the definite preferred size here would correspond to style().logicalWidth().isFixed() being true right? Those caps to the definite size would only occur in that scenario?

cathiechen replied:
IIUC, in `RenderBlock::computePreferredLogicalWidths()`:
```
m_minPreferredLogicalWidth = m_maxPreferredLogicalWidth = adjustContentBoxLogicalWidthForBoxSizing(lengthToUse);
```
It sets the fixed logicalWidth  value to `m_minPreferredLogicalWidth` and `m_maxPreferredLogicalWidth`, then it calls `RenderBox::computePreferredLogicalWidths`.
Which means, here, the constraint of `logicalWidth` has no affection. `m_minPreferredLogicalWidth` and `m_maxPreferredLogicalWidth` already have the value of  `logicalWidth`, and transferred min/max width are capped/floored by `logicalWidth`, they won't change the value of `m_minPreferredLogicalWidth` and `m_maxPreferredLogicalWidth` anyway.

sammygill replied:
This should be fixed now!

cathiechen replied:
Thanks, the patch looks good to me:)

rwlbuis commented on Source/WebCore/platform/Length.h:
```diff
@@ -231,6 +231,11 @@ inline Length& Length::operator=(Length&& other)
     return *this;
 }
 
+inline bool operator<(const Length& lhs, const Length& rhs)
```
Why is this operator needed?

sammygill replied:
It is not needed for this patch anymore. While I was working on a solution I added this since it was needed, but eventually I no longer required it. It seemed like it would be a useful operator to have since it would allow us to use std::min and std::max on Length objects. However, if you think we should remove it and only add it when it is required I can make the adjustment.

rwlbuis replied:
Yes, it is better to not include unused code in patches. Also one less thing to review! :)

sammygill replied:
Sounds good, removed!

rwlbuis reviewed (commented):
The introduced tests should probably be WPT tests.

rwlbuis reviewed (commented):
So the patch looks good but didn't you want to include additional tests?

rwlbuis commented on LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/flex-item-transferred-sizes-padding-border-sizing-expected.html:
```diff
@@ -0,0 +1,21 @@
+  <body>
+    <div></div>
+  </body>
+</html>
```
You can probably get rid of the "no newline at eof" message.

sammygill replied:
Fixed

rwlbuis commented on LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/flex-item-transferred-sizes-padding-content-sizing-expected.html:
```diff
@@ -0,0 +1,21 @@
+  <body>
+    <div></div>
+  </body>
+</html>
```
Ditto.

sammygill replied:
Fixed

rwlbuis reviewed (approved):
LGTM.

sammygill committed:
Resolve cases when border and padding are added twice to computed min and max sizes
https://bugs.webkit.org/show_bug.cgi?id=243887
rdar://98577125

Reviewed by Rob Buis.

This patch fixes a miscellaneous bug which would result in the
border and padding on boxes to get added twice in certain cases. This
is because inlineSizeFromAspectRatio returns the sizes of the border
box when the box sizing is content box. This is what is used to
compute logicalMinWidth and logicalMaxWidth, so those become
border box values. If the border and padding is not removed it is
possible for m_minPreferredLogicalWidth or m_maxPreferredLogicalWidth
to get clamped to one of those values. At the end of the method the
border and padding would once again get added on, which would result
in an incorrect box size. This could also happen in
RenderBlock::computePreferredLogicalWidths if the width is calculated
from the aspect ratio. This is because this branch of the code uses the
same inlineSizeFromAspectRatio method to do so.

Some of the logic inside of RenderBox::computePreferredLogicalWidths has
also been adjusted to no longer bother with checking transferred sizes
if a definite width is provided. This is because m_minPreferredLogicalWidth
and m_maxPreferredLogicalWidth will be set to the defined logical width
in RenderBlock::computePreferredLogicalWidths. If this is done, then
the transferred sizes will have no impact on these values. This is
because:

1.) If m_minPreferredLogicalWidth is greater than the logical width or
m_maxPreferredLogicalWidth is smaller than the logical width, they would
have been sized to the logical width anyway
2.) The clamping will have no effect if 1 does or does not occur. If 1
does occur, then m_minPreferredLogicalWidth and/or m_maxPreferredLogicalWidth
will already be the value of the logical width. If 1 does not ocur, then
m_minPreferredLogicalWidth and m_maxPreferredLogicalWidth are already
within the bounds of the transferred sizes and nothing will happen.

Transferred sizes spec: https://drafts.csswg.org/css-sizing-4/#aspect-ratio-size-transfers

* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computePreferredLogicalWidths):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computePreferredLogicalWidths):

Canonical link: https://commits.webkit.org/254190@main
